// yylib-react/dist/business/print/PrintButton.js
'use strict';

/**
 * Created by Administrator on 2016/5/11.
 * 上传附件组件
 */
var React = require('react');

var _require = require('yylib-ui');

var YYClass = _require.YYClass;
var YYButton = _require.YYButton;
var YYDropdownButton = _require.YYDropdownButton;
var YYMenu = _require.YYMenu;
var YYMenuItem = _require.YYMenuItem;

var _require2 = require("./printaction");

var loadTplList = _require2.loadTplList;
var templateUrl = _require2.templateUrl;


var PrintButton = YYClass.create({
    getDefaultProps: function getDefaultProps() {
        return { serverUrl: '', billType: '', busiType: '', orgId: '', onClickMenu: null, title: '打印', disable: false };
    },
    getInitialState: function getInitialState() {
        return { hashLoad: false, templateList: [] };
    },
    _loadList: function _loadList() {
        var url = this.props.serverUrl + templateUrl;
        var that = this;
        var billType = this.props.billType;
        var busiType = this.props.busiType;
        var orgId = this.props.orgId;
        loadTplList(url, { billType: billType, busiType: busiType, orgId: orgId }, function (data) {
            if (data) {
                that.setState({ templateList: data });
            }
        });
    },
    componentDidMount: function componentDidMount() {
        if (this.state.hashLoad == false) {
            this.setState({ hashLoad: true });
            this._loadList();
        }
    },
    handleMenuClick: function handleMenuClick(e) {
        var eventHandler = this.props.onClickMenu;
        if (eventHandler) {
            eventHandler(e.key);
        }
    },

    getMenu: function getMenu() {
        var data = this.state.templateList;
        return React.createElement(
            YYMenu,
            { onClick: this.handleMenuClick },
            data.map(function (item) {
                return React.createElement(
                    YYMenuItem,
                    { key: item.templateCode, disabled: false },
                    item.templateName
                );
            })
        );
    },
    render: function render() {
        var title = this.props.title;
        if (!title) {
            title = '打印';
        }
        var that = this;
        var menus = this.getMenu();
        return React.createElement(
            YYDropdownButton,
            { overlay: menus, type: 'primary', disabled: true },
            title
        );
    }
});

module.exports = PrintButton;
