// yylib-react/dist/yylib2/tiles/YYTilesGroup.js
'use strict';

/**
 * Created by zhangzhzhc on 2016/10/24
 * 磁贴组(同一轮播页为一组)
 */
require('./style/index.less');
var React = require('react');
var YYClass = require('../base/YYClass');
var PropTypes = YYClass.PropTypes;
var field = YYClass.field;
//定义META
var uiMeta = {
    name: 'tilesgroup',
    fields: [field('color', false, PropTypes.string, false, '背景色')]
};
var YYTilesGroup = YYClass.create({
    uiMeta: uiMeta,
    getDefaultProps: function getDefaultProps() {
        return {
            color: '#bcbcbc'
        };
    },
    getChildren: function getChildren() {
        var pre = this.props.children;
        var cur = [];
        if (pre && pre.length > 0) {
            for (var i = 0; i < pre.length; i++) {
                var rowIndex = pre[i].props.rowIndex > 0 ? pre[i].props.rowIndex : 0;
                if (!cur[rowIndex]) {
                    cur[rowIndex] = [];
                }
                pre[i].props.tilesWidth = this.props.tilesWidth;
                cur[rowIndex].push(pre[i]);
            }
            for (var j = 0; j < cur.length; j++) {
                cur[j] = React.createElement(
                    'div',
                    { className: this.className('row') },
                    ' ',
                    cur[j],
                    ' '
                );
            }
        }
        return cur;
    },
    render: function render() {
        var width = this.props.tilesWidth;
        var color = this.props.color;
        return React.createElement(
            'div',
            { className: this.className(), style: { width: width, backgroundColor: color } },
            this.getChildren()
        );
    }
});

module.exports = YYTilesGroup;
