// yylib-react/dist/business/searchfilter/YYFilterSingle.js
'use strict';

/**
 * 筛选栏-单选项组件
 */
var React = require('react');
var _ = require('lodash');
var classNames = require('classnames');

var _require = require('yylib-ui');

var YYRow = _require.YYRow;
var YYCol = _require.YYCol;


var YYFilterSingle = React.createClass({
    displayName: 'YYFilterSingle',

    getDefaultProps: function getDefaultProps() {
        return {
            showAll: false, //是否显示全部
            minCount: 8, //最少显示个数
            defaultActivedKey: '', //默认激活项
            items: [], //{value: 1, text: '显示的文字'} value不可重复
            onClick: null //点击单选时的回调函数
        };
    },
    getInitialState: function getInitialState() {
        return {
            activeKey: this.props.defaultActivedKey
        };
    },
    _handleClick: function _handleClick(item) {
        var onClick = this.props.onClick;
        var activeKey = this.state.activeKey;

        var hasSelected = activeKey === item.value ? true : false;
        if (hasSelected) {
            this.setState({ activeKey: '' });
        } else {
            this.setState({ activeKey: item.value });
        }

        if (onClick) {
            onClick(item, !hasSelected);
        }
    },
    _getItemsData: function _getItemsData(items) {
        var _props = this.props;
        var showAll = _props.showAll;
        var minCount = _props.minCount;


        return !showAll && items.length > minCount ? _.slice(items, 0, minCount) : items;
    },
    _renderContent: function _renderContent() {
        var _this = this;
        var items = this.props.items;
        var activeKey = this.state.activeKey;


        return _.map(this._getItemsData(items), function (item, index) {
            var cls = classNames({
                'yyui-filter-item': true,
                'actived': activeKey === item.value
            });
            return React.createElement(
                YYCol,
                { span: '3' },
                React.createElement(
                    'span',
                    {
                        className: cls,
                        value: item.value,
                        onClick: _this._handleClick.bind(_this, item)
                    },
                    item.text
                )
            );
        });
    },
    render: function render() {
        //判断是否展开，并显示滚动条样式
        var rowCls = classNames({
            'expand': this.props.showAll,
            'yyui-filter-single': true
        });
        return React.createElement(
            YYRow,
            { className: rowCls },
            this._renderContent()
        );
    }
});

module.exports = YYFilterSingle;
