// yylib-react/dist/yylib2/transfer/YYTransferOperation.js
'use strict';

/**
 * Created by wuhao on 16/7/18.
 */

var React = require('react');
var YYClass = require('../base/YYClass');
var field = YYClass.field;
var PropTypes = YYClass.PropTypes;
var YYIcon = require('../icon/YYIcon');
var YYButton = require('../button/YYButton');
var _ = require('lodash');
require('./style/index.less');

var uiMeta = {
    name: 'transfer-operation',
    fields: [
    //field参数列表：字段属性名|属性默认值|属性类型|是否必填|属性描述
    field('titleLeft', '左列表', PropTypes.string, false, '左边列表标题文本')]
};

var YYTransferOperation = YYClass.create({
    uiMeta: uiMeta,
    getDefaultProps: function getDefaultProps() {
        return {
            isBtnLeftDisable: false
        };
    },
    getInitialState: function getInitialState() {
        return { enableClick: true };
    },
    componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
        if ('enableClick' in nextProps) {
            if (this.state.enableClick != nextProps.enableClick) {
                this.setState({ enableClick: nextProps.enableClick });
            }
        }
    },
    _handleMove: function _handleMove() {
        var enableClick = !this.state.enableClick;
        this.setState({ enableClick: !enableClick });
        var flag = this.props.flag;
        if (_.isFunction(this.props.onHandleMove)) {
            this.props.onHandleMove(flag);
        }
    },
    render: function render() {
        var cls = 'yyui-transfer-btn-lr ' + (this.props.className ? this.props.className : '');
        var flag = this.props.flag;
        var iconTitle;
        if (flag == 'right') {
            iconTitle = React.createElement(
                'div',
                null,
                this.props.text,
                React.createElement(YYIcon, { type: flag })
            );
        } else {
            iconTitle = React.createElement(
                'div',
                null,
                React.createElement(YYIcon, { type: flag }),
                this.props.text
            );
        }

        var button;
        if (!this.props.isBtnLeftDisable) {
            button = React.createElement(
                YYButton,
                { type: 'primary', size: 'small', className: cls, style: { borderRadius: 4, border: '1px solid #ddd' },
                    disabled: this.state.enableClick, onClick: this._handleMove },
                iconTitle
            );
        }
        return React.createElement(
            'div',
            null,
            button
        );
    }
});

module.exports = YYTransferOperation;
