// yylib-react/dist/business/attach/attachaction.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.loadAttachList = loadAttachList;
exports.delAttach = delAttach;
exports.showErrMsg = showErrMsg;
/**
 * Created by ge on 2016/6/20.
 * 附件
 */

var ajax = require('../../utils/ajax');

var _require = require('yylib-ui');

var YYNotice = _require.YYNotice;
var downloadUrl = exports.downloadUrl = "/icop-file/file/download?id=";
//获取附件列表的url
var listUrl = exports.listUrl = "/icop-file/file/list";
//上传url
var uploadUrl = exports.uploadUrl = "/icop-file/file/upload2";

var delUrl = exports.delUrl = "/icop-file/file/del";

function loadAttachList(url, params, onSuccess, onError) {
    ajax.getJSON(url, params, function (data) {
        if (data.success) {
            if (onSuccess) {
                onSuccess(data.backData);
            }
        } else {
            if (onError) {
                onError("查询数据失败");
            }
        }
    }, function (error) {
        if (onError) {
            onError("服务器请求失败");
        }
    });
}

function delAttach(url, params, onSuccess, onError) {
    ajax.getJSON(url, params, function (data) {
        onSuccess(data);
    }, function (error) {
        if (onError) {
            onError("服务器请求失败");
        }
    });
}

function showErrMsg(title, msg) {
    YYNotice["error"]({
        message: title,
        description: msg
    });
}
