// yylib-react/dist/yylib2/modal/YYModal.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

/**
 * 模态对话框
 */
var React = require('react');
var YYClass = require('../base/YYClass');
var PropTypes = YYClass.PropTypes;
var field = YYClass.field;

var _require = require('antd');

var Modal = _require.Modal;
var Button = _require.Button;

require('./style/index.less');
//定义META
var uiMeta = {
    name: 'modal',
    fields: [
    //field参数列表：字段属性名|属性默认值|属性类型|是否必填|属性描述
    field('visible', null, PropTypes.bool, false, '对话框是否可见'), field('confirmLoading', null, PropTypes.bool, false, '确定按钮 loading'), field('title', null, PropTypes.element, false, '标题'), field('closable', true, PropTypes.bool, false, '是否显示右上角的关闭按钮'), field('onOk', null, PropTypes.func, false, '点击确定回调'), field('onCancel', null, PropTypes.func, false, '点击遮罩层或右上角叉或取消按钮的回调'), field('width', 520, PropTypes.number, false, '对话框宽度'), field('footer', undefined, PropTypes.element, false, '底部内容'), field('okText', '确定', PropTypes.string, false, '确认按钮文字'), field('cancelText', '取消', PropTypes.string, false, '取消按钮文字'), field('maskClosable', true, PropTypes.bool, false, '点击蒙层是否允许关闭'), field('style', {}, PropTypes.object, false, '用于设置浮层的样式，调整浮层位置'), field('wrapClassName', '', PropTypes.string, false, '对话框外层容器的类名')]
};
var YYModal = YYClass.create({
    uiMeta: uiMeta,
    statics: {
        info: Modal.info,
        success: Modal.success,
        error: Modal.error,
        warning: Modal.warning,
        confirm: Modal.confirm
    },
    handleCancel: function handleCancel(e) {
        this.props.onCancel(e);
    },
    handleOk: function handleOk() {
        this.props.onOk();
    },
    render: function render() {
        return React.createElement(
            Modal,
            _extends({}, this.props, { prefixCls: this.className() + ' ant-modal' }),
            this.props.children
        );
    }
});
module.exports = YYModal;
