// yylib-react/dist/business/quickdev/widgets/AttachMgrWidget.js
'use strict';

var React = require('react');
var _ = require('lodash');

var _require = require('yylib-ui');

var YYClass = _require.YYClass;
var YYButton = _require.YYButton;

var _require2 = require('../../attach/AttachMgr');

var showAttachMgr = _require2.showAttachMgr;


var AttachMgrWidget = YYClass.create({
    _onShowAttachMgr: function _onShowAttachMgr(config) {
        showAttachMgr(config);
    },
    render: function render() {
        var props = this.props;
        var config = {
            serverUrl: props.serverUrl,
            sourceId: props.sourceId,
            billType: props.billType,
            userId: props.userId,
            userName: props.userName,
            fileTypes: props.fileTypes,
            showUploadBtn: props.showUploadBtn,
            showDelBtn: props.showDelBtn
        };

        return React.createElement(
            YYButton,
            { onClick: this._onShowAttachMgr.bind(this, config) },
            '点击'
        );
    }
});

module.exports = AttachMgrWidget;
