// yylib-react/dist/business/enum/YYEnumSelect.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

/**
 * Created by xiejunb on 2016/12/19.
 */
var React = require('react');

var _require = require('yylib-ui');

var YYSelect = _require.YYSelect;
var YYMessage = _require.YYMessage;
var YYClass = _require.YYClass;

var field = YYClass.field;
var PropTypes = YYClass.PropTypes;
var DefDocUtil = require('yylib-utils/DefDocUtil');
var _ = require('lodash');

var uiMeta = {
    name: 'enumselect',
    fields: [
    // field参数列表：字段属性名|属性默认值|属性类型|是否必填|属性描述
    field('serverUrl', null, PropTypes.string, true, '服务url'), field('code', null, PropTypes.string, true, '自定义档案类型code'), field('orgId', null, PropTypes.string, false, '组织id'), field('valueKey', "defdocId", PropTypes.string, false, '指定下拉选项的value取值,默认取自定义档案的id,可取值defdocId、code'), field('multiple', false, PropTypes.bool, false, '支持多选'), field('allowClear', false, PropTypes.bool, false, '支持清除, 单选模式有效'), field('tags', false, PropTypes.bool, false, '可以把随意输入的条目作为 tag，输入项不需要与下拉选项匹配'), field('dropdownMatchSelectWidth', true, PropTypes.bool, false, '下拉菜单和选择器同宽'), field('combobox', false, PropTypes.bool, false, '输入框自动提示模式'), field('showSearch', false, PropTypes.bool, false, '在选择框中显示搜索框'), field('disabled', false, PropTypes.bool, false, '是否禁用'), field('defaultActiveFirstOption', true, PropTypes.bool, false, '是否默认高亮第一个选项'), field('labelInValue', false, PropTypes.bool, false, '是否把每个选项的 label 包装到 value 中，决定 Select 的 value 类型'), field('onSelect', null, PropTypes.func, false, '被选中时调用，参数为选中项的 value 值'), field('onDeselect', null, PropTypes.func, false, '取消选中时调用，参数为选中项的 option value 值，仅在 multiple 或 tags 模式下生效'), field('onChange', null, PropTypes.func, false, '选中 option，或 input 的 value 变化（combobox 模式下）时，调用此函数'), field('onSearch', null, PropTypes.func, false, '文本框值变化时回调'), field('value', undefined, PropTypes.oneOf([PropTypes.string, PropTypes.array]), false, '指定当前选中的条目'), field('defaultValue', undefined, PropTypes.oneOf([PropTypes.string, PropTypes.array]), false, '指定当前选中的条目'), field('filterOption', undefined, PropTypes.oneOf([PropTypes.bool, PropTypes.func]), false, '是否根据输入项进行筛选。当其为一个函数时，会接收 inputValue option 两个参数，当 option 符合筛选条件时，应返回 true，反之则返回 false。'), field('size', 'default', PropTypes.oneOf(['large', 'default', 'small']), false, '选择框大小，可选 large small'), field('placeholder', '', PropTypes.string, false, '选择框默认文字'), field('notFoundContent', 'Not Found', PropTypes.string, false, '当下拉列表为空时显示的内容'), field('optionFilterProp', 'value', PropTypes.string, false, '搜索时过滤对应的 option 属性，如设置为 children 表示对内嵌内容进行搜索'), field('optionLabelProp', undefined, PropTypes.string, false, '选择框回填到选择框的 Option 的属性值，默认是 Option 的子元素。比如在子元素需要高亮效果时，此值可以设为 value默认文字'), field('getPopupContainer', function () {
        return document.body;
    }, PropTypes.func, false, '菜单渲染父节点。默认渲染到 body 上，如果你遇到菜单滚动定位问题，试试修改为滚动的区域，并相对其定位')]
};
var YYEnumSelect = YYClass.create({
    uiMeta: uiMeta,
    getInitialState: function getInitialState() {
        var props = this.props;
        var value = props.value ? props.value : props.defaultValue;
        return {
            options: [],
            value: value
        };
    },
    componentDidMount: function componentDidMount() {
        this._requestData(this.props.code, this.props.orgId);
    },
    componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
        if ('defaultValue' in nextProps) {
            if (this.props.defaultValue != nextProps.defaultValue) {
                this.setState({ value: nextProps.defaultValue });
            }
        }
        //value优先级高于defaultValue
        if ('value' in nextProps) {
            if (this.props.value != nextProps.value) {
                this.setState({ value: nextProps.value });
            }
        }
        if (nextProps.code !== this.props.code) {
            this._requestData(nextProps.code, nextProps.orgId);
        }
    },
    _requestData: function _requestData(code, orgId) {
        var that = this;
        var valueKey = this.props.valueKey;
        DefDocUtil.findDefDocListByCode(this.props.serverUrl, code, orgId, function (result) {
            if (result.success) {
                var defDocList = result.backData;
                var options = defDocList.map(function (defDoc) {
                    var option = { value: defDoc[valueKey], text: defDoc.name };
                    return option;
                });
                that.setState({ options: options });
            } else {
                YYMessage.error(result.backMsg);
            }
        });
    },
    _onChange: function _onChange(keys) {
        var that = this;
        this.setState({ value: keys }, function () {
            if (_.isFunction(that.props.onChange)) {
                that.props.onChange(keys);
            }
        });
    },
    render: function render() {
        return React.createElement(YYSelect, _extends({}, this.props, { items: this.state.options,
            onChange: this._onChange, value: this.state.value }));
    }
});

module.exports = YYEnumSelect;
