// yylib-react/dist/yylib2/formcontrols/cascader/YYCascader.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var React = require('react');
var YYClass = require('../../base/YYClass');

var _require = require('react-dom');

var findDOMNode = _require.findDOMNode;

var PropTypes = YYClass.PropTypes;
var field = YYClass.field;

var _require2 = require('antd');

var Cascader = _require2.Cascader;

var _ = require('lodash');
var ajax = require('../../../utils/ajax');
//定义META
var uiMeta = {
    name: 'cascader',
    fields: [
    // field参数列表：字段属性名|属性默认值|属性类型|是否必填|属性描述
    field('options', null, PropTypes.object, false, '本地数据项'), field('dataurl', null, PropTypes.string, false, '请求远程数据源的URL，注意当配置了dataurl时，options属性将失效'), field('dataparams', null, PropTypes.object, false, '请求远程数据时的自定义参数'), field('defaultValue', undefined, PropTypes.array, false, '默认的选中项'), field('value', undefined, PropTypes.array, false, '指定选中项'), field('style', {}, PropTypes.object, false, '自定义样式'), field('className', '', PropTypes.string, false, '自定义类名'), field('popupClassName', '', PropTypes.string, false, '自定义浮层类名'), field('placeholder', '请选择', PropTypes.string, false, '输入框占位文本'), field('size', 'default', PropTypes.oneOf(['large', 'default', 'small']), false, '输入框大小'), field('disabled', false, PropTypes.bool, false, '禁用'), field('allowClear', true, PropTypes.bool, false, '是否支持清除'), field('changeOnSelect', false, PropTypes.bool, false, '当此项为 true 时，点选每级菜单选项值都会发生变化，具体见上面的演示'), field('popupPlacement', 'bottomLeft', PropTypes.oneOf(['bottomLeft', 'bottomRight', 'topLeft', 'topRight']), false, '浮层预设位置'), field('expandTrigger', 'click', PropTypes.oneOf(['click', 'hover']), false, '次级菜单的展开方式'), field('onChange', null, PropTypes.func, false, '选择完成后的回调'), field('displayRender', null, PropTypes.func, false, '选择后展示的渲染函数')]
};

var YYCascader = YYClass.create({
    uiMeta: uiMeta,
    getInitialState: function getInitialState() {
        return {
            options: []
        };
    },
    componentDidMount: function componentDidMount() {
        var _props = this.props;
        var dataurl = _props.dataurl;
        var dataparams = _props.dataparams;

        this._requestData(dataurl, dataparams);
    },
    componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
        var _props2 = this.props;
        var dataurl = _props2.dataurl;
        var dataparams = _props2.dataparams;

        if (nextProps.dataurl !== dataurl || !_.isEqual(nextProps.dataparams, dataparams)) {
            this._requestData(nextProps.dataurl, nextProps.dataparams);
        }
    },
    _requestData: function _requestData(dataurl, dataparams) {
        var _this = this;
        if (dataurl) {
            if (dataparams) {
                ajax.getJSON(dataurl, dataparams, function (result) {
                    _this.setState({ options: result });
                });
            } else {
                ajax.getJSON(dataurl, function (result) {
                    _this.setState({ options: result });
                });
            }
        }
    },
    handleKeyDown: function handleKeyDown(e) {
        var code = e.keyCode;
        var cascader = this.refs.cascader;
        if (code == 13) {
            cascader.setState({
                popupVisible: true
            });
        };
    },
    render: function render() {
        var options = this.props.dataurl ? this.state.options : this.props.options;
        return React.createElement(
            'div',
            { className: this.className(), onKeyDown: this.handleKeyDown },
            React.createElement(
                Cascader,
                _extends({}, this.props, { options: options, ref: 'cascader' }),
                this.props.children
            )
        );
    }
});

module.exports = YYCascader;
