// yylib-react/dist/yylib2/refer/YYReferGrid.js
'use strict';

var _YYReferOperateDlg = require('./YYReferOperateDlg');

var _YYReferOperateDlg2 = _interopRequireDefault(_YYReferOperateDlg);

var _ReferConstants = require('./constants/ReferConstants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var React = require('react');
var _ = require('lodash');
var ajax = require('../../utils/ajax');
var YYClass = require('../base/YYClass');
var YYTabs = require('../tabs/YYTabs');
var YYTab = require('../tabs/YYTab');
var YYIcon = require('../icon/YYIcon');
var YYRow = require('../layout/YYRow');
var YYCol = require('../layout/YYCol');
var YYButton = require('../button/YYButton');
var YYTable = require('../table/YYTable');
var YYPagination = require('../paging/YYPagination');
var YYSelect = require('../formcontrols/select/YYSelect');
var YYMessage = require('../message/YYMessage');
var YYScrollbar = require('../scrollbar/YYScrollbar');
var YYInputButton = require('../formcontrols/inputbutton/YYInputButton');
var YYFootInRefer = require('./YYFootInRefer');
var PropTypes = YYClass.PropTypes;
var field = YYClass.field;

var _require = require('react-redux');

var connect = _require.connect;

var _require2 = require('./actions/reference');

var getGridData = _require2.getGridData;
var addOften = _require2.addOften;
var deleteOften = _require2.deleteOften;

require('./style/index.less');
var AuthToken = require("../../utils/AuthToken");
// var pageSize = 10;
var store = window.localStorage;
var tableKey = null;
var charCode = String.fromCharCode(2);

var _require3 = require('./utils/YYReferUtils');

var getColumns = _require3.getColumns;
var ifInArray = _require3.ifInArray;
var removeRow = _require3.removeRow;
var getPagination = _require3.getPagination;
var getDefaultRows = _require3.getDefaultRows;
var getDefaultKeys = _require3.getDefaultKeys;
var getHeaders = _require3.getHeaders;
var formatOftenToServer = _require3.formatOftenToServer;
var formatOftenFromServer = _require3.formatOftenFromServer;
var getOftenKeys = _require3.getOftenKeys;

var mapStateToProps = function mapStateToProps(state) {
	return {
		refinfo: state.refinfo
	};
};

function noop() {};
//定义META
var uiMeta = {
	name: 'refer-grid',
	fields: [
	//field参数列表：字段属性名|属性默认值|属性类型|是否必填|属性描述
	field('refinfokey', null, PropTypes.string, true, '参照key'), field('refname', null, PropTypes.string, true, '参照名称'), field('dataurl', null, PropTypes.string, true, '查询url'), field('operateurl', null, PropTypes.string, true, '操作url'), field('initialValue', null, PropTypes.oneOfType([PropTypes.object, PropTypes.Array]), false, '默认初始值'), field('orderfield', null, PropTypes.string, true, '排序字段'), field('often', null, PropTypes.bool, true, '是否显示常用'), field('multiselect', null, PropTypes.bool, true, '是否支持多选'), field('querystr', null, PropTypes.string, true, '查询条件(过期)'), field('condition', null, PropTypes.oneOfType([PropTypes.func, PropTypes.object]), true, '自定义查询条件'), field('onOk', null, PropTypes.func, true, '确认操作回调函数'), field('onCancel', null, PropTypes.func, true, '取消操作回调函数'), field('customheader', null, PropTypes.element, false, '自定义header'), field('contentH', null, PropTypes.number, true, '内容区域的高度'), field('gridheaders', null, PropTypes.array, false, '表头信息')]
};
var YYReferGrid = YYClass.create({
	uiMeta: uiMeta,
	getInitialState: function getInitialState() {
		return {
			current: 1,
			refid: null,
			name: null,
			loadOften: false,
			selectedRow: this.props.multiselect ? null : getDefaultRows(this.props.initialValue, this.props.idField, this.props.codeField, this.props.nameField, this.props.multiselect), //单选row
			selectedRowKey: this.props.multiselect ? null : getDefaultKeys(this.props.initialValue, this.props.multiselect), //单选key,多选模式下不获取
			selectedRows: this.props.multiselect ? getDefaultRows(this.props.initialValue, this.props.idField, this.props.codeField, this.props.nameField, this.props.multiselect) : null, //多选rows
			selectedRowKeys: this.props.multiselect ? getDefaultKeys(this.props.initialValue, this.props.multiselect) : null, //多选keys, 单选模式下不获取
			searchVal: null,
			loading: false,
			showOperateDialog: false, //是否打开operate dialog
			operateurl: null };
	},
	componentWillMount: function componentWillMount() {
		var refinfokey = this.props.refinfokey;

		if (store) {
			var userCode = AuthToken.getUserCode();
			tableKey = _ReferConstants.REFER_TABLEKEY + userCode + "-" + refinfokey;
		}
	},
	componentDidMount: function componentDidMount() {
		var that = this;
		// 延迟，解决tab错位的bug
		setTimeout(function () {
			that.doQuery(null, 1);
		}, 500);
	},
	componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
		this.setState({ loading: false });
		if (nextProps && nextProps.refinfo) {
			var refinfo = nextProps.refinfo;
			if (refinfo.doReload) {
				//重新查询数据
				this.doQuery(null, 1);
				this.refs.inputSearch.state.value = '';
				this.setState({
					current: 1
				});
			}
			if (refinfo.showMsg) {
				var showMsg = refinfo.showMsg;
				if (showMsg.success) {
					YYMessage.success(showMsg.success);
				} else if (showMsg.error) {
					YYMessage.error(showMsg.error);
				}
			}
		}
	},
	changeTags: function changeTags(key) {
		this.setState({
			loadOften: !this.state.loadOften
		});
	},

	// 保存/删除 常用数据到服务器. oper=0: 手动， oper=1：自动
	_saveOften: function _saveOften(records, oper, isDelete) {
		var dispatch = this.props.dispatch;
		var userId = AuthToken.getUserId();
		var refinfokey = this.props.refinfokey;
		var idField = this.props.idField;
		var serverUrl = this.props.serverUrl;
		if (isDelete === true) {
			var ids = records[idField]; // 当前删除只支持单个删除，后续支持批量删除后在此扩展
			dispatch(deleteOften(serverUrl, { userId: userId, refinfokey: refinfokey, ids: ids }));
		} else {
			var dataset = formatOftenToServer(records, idField);
			if (dataset && dataset.length > 0) {
				dispatch(addOften(serverUrl, { userId: userId, refinfokey: refinfokey, oper: oper, dataset: dataset }));
			}
		}
	},
	addOften: function addOften(record, e) {
		e.stopPropagation();
		if (store && record) {
			var nodeKey = record[this.props.idField];
			this._saveOften(record, 0);
		}
		this.setState({
			loadOften: !this.state.loadOften
		});
	},
	delOften: function delOften(record, e) {
		e.stopPropagation();
		var nodeKey = record[this.props.idField];
		var isDelete = true; // 便于理解和维护。此处用于标识删除操作
		this._saveOften(record, null, isDelete);
		this.setState({
			loadOften: !this.state.loadOften
		});
	},
	onCancelHandle: function onCancelHandle(e) {
		if (this.props.onCancel) {
			this.props.onCancel(e);
		}
	},

	// 清除已选择的数据
	onClean: function onClean() {
		this.setState({
			selectedRow: null,
			selectedRowKey: null,
			selectedRows: null,
			selectedRowKeys: null
		});
	},

	//设置查询条件
	getQuery: function getQuery(querystr, condition, searchText, orderfield, pageNumber, pageSize) {
		var getQuery = {};
		getQuery.pageNumber = pageNumber;
		getQuery.pageSize = pageSize ? pageSize : YYTable.getDefaultPageSize(tableKey);
		if (querystr && querystr.length > 0) getQuery.relyCondition = querystr;
		if (orderfield && orderfield.length > 0) getQuery.orderCondition = orderfield;
		if (searchText && searchText.length > 0) getQuery.searchText = searchText;
		var _condition = ajax.getParams(condition);
		if (!_.isEmpty(_condition) && _.isPlainObject(_condition)) {
			//转成JSON格式字符串
			getQuery.condition = JSON.stringify(_condition);
		}
		return getQuery;
	},

	//执行查询动作
	doQuery: function doQuery(searchVal, current, pageSize) {
		var _props = this.props;
		var dataurl = _props.dataurl;
		var querystr = _props.querystr;
		var condition = _props.condition;
		var orderfield = _props.orderfield;
		var refid = _props.refid;
		var dispatch = _props.dispatch;

		pageSize = pageSize ? pageSize : YYTable.getDefaultPageSize(tableKey);
		var getQuery = this.getQuery(querystr, condition, searchVal, orderfield, current, pageSize);
		this.setState({ loading: true });
		dispatch(getGridData(dataurl, getQuery));
	},

	//页码更改
	onChangePage: function onChangePage(current) {
		this.doQuery(this.refs.inputSearch.state.value, current);
		this.setState({
			current: current
		});
	},

	//每页显示的条数更改
	onChangePageSize: function onChangePageSize(current, pageSize) {
		this.doQuery(this.refs.inputSearch.state.value, current, pageSize);
		this.setState({
			current: current
		});
	},

	//搜索
	onSearch: function onSearch(e) {
		var value = this.refs.inputSearch.state.value;
		if (this.state.loading) {
			return;
		} else {
			this.doQuery(value, 1);
			this.setState({
				current: 1,
				searchVal: value
			});
		}
	},
	doHandleDatas: function doHandleDatas(selectedRows, selectedRow, isHandleOk) {
		var multiselect = this.props.multiselect;

		var row = null;
		if (multiselect) {
			row = selectedRows ? selectedRows : this.state.selectedRows;
		} else {
			row = selectedRow ? selectedRow : this.state.selectedRow;
		}
		if (isHandleOk) {
			if (this.props.often) {
				this._saveOften(row, 1);
			}
			this.props.onOk(row);
		} else {
			this.props.onSelectedChanged(row);
		}
	},
	onOkHandle: function onOkHandle(e) {
		this.doHandleDatas(null, null, true);
	},
	onChangeRow: function onChangeRow(keys, rows) {
		var selectedRows = this.state.selectedRows ? this.state.selectedRows : [];
		var selectedRowKeys = this.state.selectedRowKeys ? this.state.selectedRowKeys : [];
		if (rows) {
			for (var i = rows.length - 1; i >= 0; i--) {
				if (!ifInArray(selectedRowKeys, rows[i][this.props.idField])) {
					selectedRowKeys.push(rows[i][this.props.idField]);
					selectedRows.push(rows[i]);
				};
			}
		} else {
			selectedRows = null;
		}
		this.setState({
			selectedRows: selectedRows,
			selectedRowKeys: selectedRowKeys
		});
		this.doHandleDatas(selectedRows, null, false);
	},
	onSelectRow: function onSelectRow(record, selected, selectRows) {
		var selectedRows = this.state.selectedRows ? this.state.selectedRows : [];
		var selectedRowKeys = this.state.selectedRowKeys;
		if (!selected) {
			var key = record[this.props.idField];
			selectedRowKeys.indexOf(key) > -1 ? selectedRowKeys.splice(selectedRowKeys.indexOf(key), 1) : null;
			selectedRows = removeRow(selectedRows, key, this.props.idField);
			this.setState({
				selectedRows: selectedRows,
				selectedRowKeys: selectedRowKeys
			});
			this.doHandleDatas(selectedRows, null, false);
		}
	},
	onSelectAllRow: function onSelectAllRow(selected, selectRows, changeRows) {
		var selectedRows = this.state.selectedRows ? this.state.selectedRows : [];
		var selectedRowKeys = this.state.selectedRowKeys;
		if (!selected) {
			for (var i = changeRows.length - 1; i >= 0; i--) {
				var key = changeRows[i][this.props.idField];
				selectedRowKeys.indexOf(key) > -1 ? selectedRowKeys.splice(selectedRowKeys.indexOf(key), 1) : null;
				selectedRows = removeRow(selectedRows, key, this.props.idField);
			};
			this.setState({
				selectedRows: selectedRows,
				selectedRowKeys: selectedRowKeys
			});
			this.doHandleDatas(selectedRows, null, false);
		}
	},

	//选中行 单选时
	onRowClick: function onRowClick(record, index) {
		this.setState({
			selectedRow: record,
			selectedRowKey: record[this.props.idField]
		});
		this.doHandleDatas(null, record, false);
	},

	// 选中行 双击时 单选时
	onRowDoubleClick: function onRowDoubleClick(record, index) {
		this.doHandleDatas(null, record, true);
	},
	onTagClose: function onTagClose(key) {
		var selectedRows = this.state.selectedRows;
		var selectedRowKeys = this.state.selectedRowKeys;
		selectedRowKeys.indexOf(key) > -1 ? selectedRowKeys.splice(selectedRowKeys.indexOf(key), 1) : null;
		removeRow(selectedRows, key, this.props.idField);
		this.setState({
			selectedRows: selectedRows,
			selectedRowKeys: selectedRowKeys
		});
		this.doHandleDatas(selectedRows, null, false);
	},

	//点击带链接字段
	onClickOperate: function onClickOperate(record, e) {
		if (record) {
			var url = this.props.operateurl + record[this.props.idField];
			this.setState({
				showOperateDialog: true,
				operateurl: url
			});
		}
	},

	//关闭点击链接打开的dialog回调
	onCloseOperate: function onCloseOperate() {
		this.setState({
			showOperateDialog: false
		});
	},

	//渲染标签页
	_renderTabs: function _renderTabs() {
		var _this = this;

		var that = this;
		var _props2 = this.props;
		var refinfo = _props2.refinfo;
		var refname = _props2.refname;
		var multiselect = _props2.multiselect;

		var grid = refinfo.grid;
		var header = getHeaders(that.props.gridheaders, grid);

		var columnsInStore = getColumns(header, store, this.props.nameField, this.props.operatefield, this.props.operateurl, this.onClickOperate);
		var totalWidth = columnsInStore.totalWidth;

		var griddata = grid ? grid.data : null;
		var totalsize = griddata ? griddata.count : 0; //总数量
		var content = griddata ? griddata.content : null;
		var data = [];
		if (content && content.length > 0) {
			data = content;
			data.map(function (elem, eindex) {
				elem.key = elem[_this.props.idField];
			});
		};
		var rowSelection = null;
		if (multiselect) {
			rowSelection = {
				onChange: this.onChangeRow,
				onSelect: this.onSelectRow,
				onSelectAll: this.onSelectAllRow,
				selectedRowKeys: this.state.selectedRowKeys
			};

			totalWidth += _ReferConstants.REFER_CHECKBOX_WIDTH;
		};
		var tableH = this.props.contentH - 28 - 45 - 10; //28是表格上方搜索框的高度、45是分页的高度，10是marginTop
		var pagination = getPagination(totalsize, that.state.current, that.onChangePage, that.onChangePageSize);

		var columns = columnsInStore.columns;
		var oftenTab = null;
		if (this.props.often) {
			var oftenData = formatOftenFromServer(this.props.refinfo.often);
			var oftenkeys = getOftenKeys(this.props.refinfo.often);
			var oftenColumns = columnsInStore.oftenColumns;
			columns.push({
				key: 'operation',
				title: _ReferConstants.REFER_ADD_OFTEN,
				width: _ReferConstants.REFER_OPERATE_COLUMN_WIDTH,
				render: function render(text, record) {
					return ifInArray(oftenkeys, record[_this.props.idField]) ? React.createElement(YYButton, { icon: 'minus', ghost: true, onClick: _this.delOften.bind(null, record) }) : React.createElement(YYButton, { icon: 'plus', ghost: true, onClick: _this.addOften.bind(null, record) });
				}
			});
			oftenColumns.push({
				key: 'operationOften',
				title: _ReferConstants.REFER_DELETE_OFTEN,
				width: _ReferConstants.REFER_OPERATE_COLUMN_WIDTH,
				render: function render(text, record) {
					return React.createElement(YYButton, { icon: 'minus', ghost: true, onClick: _this.delOften.bind(null, record) });
				}
			});

			totalWidth += _ReferConstants.REFER_OPERATE_COLUMN_WIDTH;
			oftenTab = React.createElement(
				YYTab,
				{ key: refname + "_often", tab: '常用', disabled: !this.props.often, className: 'yyui-refer-tab' },
				React.createElement(YYTable, {
					style: { marginTop: 5 },
					columns: oftenColumns,
					pagination: false,
					rowKey: that.props.idField,
					dataSource: oftenData,
					scroll: { x: totalWidth, y: this.props.contentH } //table下分页的高度
					, rowActiveKey: that.state.selectedRowKey,
					rowSelection: rowSelection,
					onRowDoubleClick: multiselect ? noop : this.onRowDoubleClick,
					onRowClick: multiselect ? noop : this.onRowClick })
			);
		}
		var dataTab = React.createElement(
			YYTab,
			{ key: refname + "_data", tab: refname, className: 'yyui-refer-tab' },
			this.props.customheader,
			React.createElement(
				'div',
				{ className: 'yyui-refer-grid-search' },
				React.createElement(YYInputButton, { ref: 'inputSearch', placeholder: "请输入搜索的" + refname, buttonIcon: 'search', onPressEnter: this.onSearch, buttonClick: this.onSearch })
			),
			React.createElement(YYTable, {
				style: { marginTop: 5 },
				cookieKey: tableKey,
				columns: columns,
				loading: this.state.loading,
				pagination: pagination,
				rowKey: that.props.idField,
				dataSource: data,
				scroll: { x: totalWidth, y: tableH - _ReferConstants.REFER_TABLE_PAGINATION_HEIGHT },
				rowActiveKey: that.state.selectedRowKey,
				onRowDoubleClick: multiselect ? noop : this.onRowDoubleClick,
				onRowClick: multiselect ? noop : this.onRowClick,
				rowSelection: rowSelection })
		);
		if (oftenTab !== null) {
			return React.createElement(
				YYTabs,
				{ defaultActiveKey: refname + "_data", onChange: this.changeTags },
				oftenTab,
				dataTab
			);
		} else {
			return React.createElement(
				YYTabs,
				{ defaultActiveKey: refname + "_data", onChange: this.changeTags },
				dataTab
			);
		}
	},
	render: function render() {
		var that = this;
		var props = {};
		props.refname = this.props.refname;
		props.multiselect = this.props.multiselect;
		props.idField = this.props.idField;
		props.nameField = this.props.nameField;
		props.codeField = this.props.codeField;
		props.onOk = this.onOkHandle;
		props.onCancel = this.onCancelHandle;
		props.onClean = this.onClean;
		props.onTagClose = this.onTagClose;
		props.selectedDatas = this.state.selectedRows;
		props.refinfokey = this.props.refinfokey;
		return React.createElement(
			'div',
			{ className: this.className() },
			this._renderTabs(),
			React.createElement(_YYReferOperateDlg2.default, { show: that.state.showOperateDialog, operateurl: that.state.operateurl, onCancel: that.onCloseOperate }),
			React.createElement(YYFootInRefer, props)
		);
	}
});
module.exports = connect(mapStateToProps)(YYReferGrid);
